lib/generators/aws_record/base.rb

```diff
@@ -0,0 +1,218 @@
+require 'rails/generators'
+require_relative 'generated_attribute'
+require_relative 'secondary_index'
+
+module AwsRecord
+    module Generators
+      class Base < Rails::Generators::NamedBase
+        argument :attributes, type: :array, default: [], banner: "field[:type][:opts]...", desc: "Describes the fields in the model"
+        check_class_collision
+
+        class_option :disable_mutation_tracking, type: :boolean, desc: "Disables dirty tracking"
+        class_option :timestamps, type: :boolean, desc: "Adds created, updated timestamps to the model"
+        class_option :table_config, type: :hash, default: {}, banner: "primary:R-W [SecondaryIndex1:R-W]...", desc: "Declares the r/w units for the model as well as any secondary indexes", :required => true
+        class_option :gsi, type: :array, default: [], banner: "name:hkey{field_name}[,rkey{field_name},proj_type{ALL|KEYS_ONLY|INCLUDE}]...", desc: "Allows for the declaration of secondary indexes"
+        class_option :table_name, type: :string, banner: "model_table_name"
+        class_option :password_digest, type: :boolean, desc: "Whether to add a password_digest field to the model"
+
+        class_option :required, type: :string, banner: "field1...", desc: "A list of attributes that are required for an instance of the model"
+        class_option :length_validations, type: :hash, default: {}, banner: "field1:MIN-MAX...", desc: "Validations on the length of attributes in a model"
+        class_option :scaffold, type: :boolean, desc: "Adds helper methods that scaffolding uses"
+
+        attr_accessor :primary_read_units, :primary_write_units, :gsi_rw_units, :gsis, :required_attrs, :length_validations
+
+        private
+
+        def initialize(args, *options)
+          options[0] << "--skip-table-config" if options[1][:behavior] == :revoke
+          @parse_errors = []
+
+          super
+          ensure_unique_fields
+          ensure_hkey
+          parse_gsis!
```

Comment:
I'm not sure why some of these methods are `!` and others are not.  I'd argue for consistency, just leave them out and make all of them private.